### ISSUE TITLE
Switch on interface with single case not considered exhaustive

### ISSUE THREAD

**Issue by mpal9000** [Microsoft/TypeScript#18851]
**TypeScript Version:**  2.5.2

**Code**

```ts
interface Action {
  readonly type: 'Action1',
}

export const assertNever = (action:never):never => {
  throw new Error('Unhandled action')
}

const fn = (action: Action) => {
  switch (action.type) {
    case 'Action1': {
      return 'action1'
    }

    default: {
      return assertNever(action) // incorrect error
    }
  }
}
```

**Expected behavior:**
`action` inside `default` block should have type `never`

**Actual behavior:**
`action` inside `default` block has type `Action`

Similar fixed issue: #12771

Playground: https://goo.gl/rs6TgW

**Comment by DanielRosenwasser** [comment]
Hey @mpal9000, this looks to me like a potential duplicate of #18056

**Comment by mpal9000** [comment]
Hi, you can close this one if you want. Checking `action.type` in `assertNever()` is not a solution, because it doesn't work for unions. A built in exhaustiveness checking in `switch` and future `match`, without manual assertions, would be an awesome feature in my opinion.
